contracts/RemovalQueue.sol

```diff
@@ -0,0 +1,177 @@
+            break;
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * @notice Checks if the queue is empty across all vintages.
+   * @dev Uses the latestYear property to check if any vintages have been set.
+   * @param removalQueue the queue from storage.
+   * @return bool true if empty, false otherwise.
+   */
+  function isRemovalQueueEmpty(RemovalQueueByVintage storage removalQueue)
+    internal
+    view
+    returns (bool)
+  {
+    return removalQueue.latestYear == _DEFAULT_LATEST_YEAR;
+  }
+
+  /**
+   * @notice Checks if the queue is empty for a particular vintage.
+   * @param removalQueue the queue from storage.
+   * @param vintage the vintage to check.
+   * @return bool true if empty, false otherwise.
+   */
+  function isRemovalQueueEmptyForVintage(
+    RemovalQueueByVintage storage removalQueue,
+    uint16 vintage
+  ) internal view returns (bool) {
+    return getSizeOfQueueForVintage(removalQueue, vintage) == 0;
+  }
+
+  /**
+   * @notice Gets the next removal in the queue for sale.
+   * @dev Gets the first item from the Enumerable Set that corresponds to the earliest vintage.
+   * @param removalQueue the queue from storage.
+   * @return uint256 the removal for sale.
+   */
+  function getNextRemovalForSale(RemovalQueueByVintage storage removalQueue)
+    internal
+    view
+    returns (uint256)
+  {
+    return removalQueue.queueByVintage[removalQueue.earliestYear].at(0);
+  }
+
+  /**
+   * @notice Gets the size of the queue for a particular vintage.
+   * @dev Gets the size of the Enumerable Set that corresponds to the given vintage.
+   * @param removalQueue the queue from storage.
+   * @param vintage the vintage to check.
+   * @return uint256 the size of the queue.
+   */
+  function getSizeOfQueueForVintage(
+    RemovalQueueByVintage storage removalQueue,
+    uint16 vintage
+  ) internal view returns (uint256) {
+    return removalQueue.queueByVintage[vintage].length();
+  }
+
+  /**
+   * @notice Gets the total balance of all removals across all vintages.
+   * @dev Gets the size of the Enumerable Set that corresponds to the given vintage.
+   * @param removalQueue the queue from storage.
+   * @param removal the removal contract.
+   * @return uint256 the total balance of the queue.
+   */
+  function getTotalBalanceFromRemovalQueue(
+    RemovalQueueByVintage storage removalQueue,
+    Removal removal
+  ) internal view returns (uint256) {
+    uint256 size = 0;
+    uint256 i = 0;
+    uint256 totalBalance = 0;
+    uint256 currentRemoval;
+    for (
+      uint16 currentYear = removalQueue.earliestYear;
+      currentYear <= removalQueue.latestYear;
+      currentYear++
+    ) {
+      size = removalQueue.queueByVintage[currentYear].length();
+      for (i = 0; i < size; i++) {
+        currentRemoval = removalQueue.queueByVintage[currentYear].at(i);
+        totalBalance += removal.balanceOf(address(this), currentRemoval); // TODO: Use batch get for balanceOf
+      }
+    }
```

Comment:
is there a reason this can't be done with something like:
```
balanceOfBatch(...address(this)[], supplierSet[supplierAddress].values()) // pseudo code, but should convey the point
```

Reply:
You need to get the extract the values from each enumerable set within the mapping before you could do this, so it's not as simple as what you've written.

Reply:
there is significant overhead for each time you invoke `.balanceOf` (vs a single balanceOfBatch) so I'd like to see this extracted to a batch version

Reply:
Agree, one way or another I'll get this into a balanceOfBatch